src/Workspaces/Core/Portable/Workspace/Workspace.cs

```diff
@@ -380,16 +394,19 @@ protected internal void OnSolutionAdded(SolutionInfo solutionInfo)
         {
             using (_serializationLock.DisposableWait())
             {
-                var oldSolution = this.CurrentSolution;
                 var solutionId = solutionInfo.Id;
 
                 CheckSolutionIsEmpty();
-                this.SetCurrentSolution(this.CreateSolution(solutionInfo));
 
-                solutionInfo.Projects.Do(p => OnProjectAdded_NoLock(p, silent: true));
+                var (oldSolution, newSolution) = this.SetCurrentSolution(this.CreateSolution(solutionInfo));
 
-                var newSolution = this.CurrentSolution;
-                this.RaiseWorkspaceChangedEventAsync(WorkspaceChangeKind.SolutionAdded, oldSolution, newSolution);
+                // Pass 'silent: true' so we don't issue individual project events.  We'll just issue one event
+                // representing the entire solution-add.
+                foreach (var project in solutionInfo.Projects)
+                    (_, newSolution) = OnProjectAdded_NoLock(project, silent: true);
```

Comment:
i like this pattern as it makes it explicit htat we're not looking at or updating 'oldSolutoin' here (since we want the very first value we got back in SetCurrentSolution above.